### ISSUE TITLE
Glitch in displaying item labels

### ISSUE THREAD

**Issue by Reiquard** [jptrrs/HumanResources#128]
Under certain conditions there is a glitch in displaying item labels.

![screenshot33](https://user-images.githubusercontent.com/62316596/106369354-c9969a80-6361-11eb-9db2-09b47130edb2.png)

This can be reproduced with the following conditions:
 - Newly started RimWorld
 - Just loaded save with installed study desk and an unfinished book on it
 - Click on any pawn and **only then** click on this book (if you click on a book first, its label is truncated, as expected)

Until the next RimWorld restart, this label will remain un-truncated (even after reloading a save)

This glitch isn't limited to just unfinished books, it can be seen on unfinished sculpture, for example.
Moreover, with similar actions (click a pawn and then click on any item that you haven't clicked on previously), you can see this glitch on other things with long enough names (not so common in English RW, so I will give an example in my own language)

![screenshot34](https://user-images.githubusercontent.com/62316596/106369359-d1563f00-6361-11eb-9443-06c1309f3850.png)

activeMods
      - brrainz.harmony
      - ludeon.rimworld
      - ludeon.rimworld.royalty
      - unlimitedhugs.hugslib
      - fluffy.modmanager
      - vinalx.researchpalforked
      - jpt.humanresources
/activeMods

**Comment by jptrrs** [comment]
> This glitch isn't limited to just unfinished books, it can be seen on unfinished sculpture, for example.

Just one question: does it happen in vanilla too? Or is it just when HR is loaded?

**Comment by Reiquard** [comment]
I have checked multiple times and could not reproduce this glitch in (vanilla):

activeMods
      - brrainz.harmony
      - ludeon.rimworld
      - ludeon.rimworld.royalty
      - unlimitedhugs.hugslib
      - fluffy.modmanager
      - vinalx.researchpalforked
/activeMods

**Comment by Reiquard** [comment]
I just thought that all the items in the examples I gave are stuffed (hence the names of these things are lengthened).

I have an assumption that labels of non-stuffed things will be truncated as expected given that their names are long enough for this (I can't check my guess right now, unfortunately).

So I believe that the reason for this may lie in the "[CreateStuff](https://github.com/jptrrs/HumanResources/blob/1ff43c0f978f31efd12a2babd790ccc16154d29a/Source/Extensions/Extension_Research.cs#L139)" method, its caller "[DefsLoaded](https://github.com/jptrrs/HumanResources/blob/1ff43c0f978f31efd12a2babd790ccc16154d29a/Source/Base/ModBaseHumanResources.cs#L58)" or somewhere near.

**EDIT:** _I just checked - this bug persists in simple, non-stuffed items like steel as well._

Latest findings:
 - Clicking on a dog, alpaca, etc. doesn't trigger this to happen. Only colonist, friendly or enemy (humanoid?) pawn does.
 - Lets say we have three research benches: two wooden (B1 and B2) and one steel (B3).
After clicking on a pawn and then on a B1, its name (as well as B2's) will be displayed with that glitch; and when we select B3 then, its name will get truncated. **So this is the oddity not of the thing itself, but the def's label.**
 - If I click on (select) a pawn, then deselect it and only then click on a thing, its name get truncated. **So this glitch happens only when selection changed from a pawn to the long named item.**

Here is a clue:
_Commenting [lines 70-83](https://github.com/jptrrs/HumanResources/blob/1ff43c0f978f31efd12a2babd790ccc16154d29a/Source/Base/ModBaseHumanResources.cs#L70-L83) gets rid of that glitch (although also removes the Tech-tab from a pawn)_

**/EDIT**

**Comment by Reiquard** [comment]
I tried to approach the problem from a different angle..

**RimWorld.MainTabWindow_Inspect.GetLabel**(Rect rect) => 
**InspectPaneUtility.AdjustedLabelFor**(Selected, rect) => 
**text.Truncate**(rect.width, truncatedLabelsCached)

```
public static string Truncate(this string str, float width, Dictionary<string, string> cache = null)
{
	if (cache != null && cache.TryGetValue(str, out string value))
	{
		return value;
	}
	if (Text.CalcSize(str).x <= width)
	{
		cache?.Add(str, str);
		return str;
	}
	value = str;
	do
	{
		value = value.Substring(0, value.Length - 1);
	}
	while (value.Length > 0 && Text.CalcSize(value + "...").x > width);
	value += "...";
	cache?.Add(str, value);
	return value;
}
```
If I get it right, when we select an item for the first time, its name is truncated and the truncation result is cached, and the next time we select this item, its name no longer goes through the truncation procedure as long as the cached result exists.

It explains why this glitch persists throughout the game session, but the question is: how could the long name have been cached?

**Comment by jptrrs** [comment]
This is probably the strangest bug I've chased so far. Luckly it's pretty harmless too. ;-) Thanks for identifying and isolating it. It looks as something most people, including me, wouldn't even notice, let alone pursue the offending mod!
I had to change the label for a vanilla object and make it longer to test it out. And I can confirm that, indeed, this can affect any unfinished item, not only books. It might even affect regular items. 
It seems to be related to the adding of the Tech tab to every human-like pawn. That explains why it doesn't happen if you click on animals before selecting the unfinished thing. Once a pawn's inspector panel is opened, it seems to disrupt the truncated title generation somehow, and since that cache is created only when the panel is first instantiated (via the InspectPaneUtility constructor), it will depend on what type of pawn was selected first on that particular game session.
I'l try to look for a way to prevent this, but I'll make no promises. It's something pretty deep into the game UI territory and I'm not sure I want to tamper with that more than I already did...
Let me know if you find anything else about how adding a tab to the pawns is able to disrupt the cache generation.
Thanks!

**Comment by jptrrs** [comment]
For the record, since I borrowed the pawn tab injection method from [Psychology](https://steamcommunity.com/sharedfiles/filedetails/?id=2016263135), I had to test that mod too. And, lo and behold, the same thing happens!

**Comment by Reiquard** [comment]
> Once a pawn's inspector panel is opened, it seems to disrupt the truncated title generation somehow, and since that cache is created only when the panel is first instantiated (via the InspectPaneUtility constructor), it will depend on what type of pawn was selected first on that particular game session.

That's not quite right. It would be more correct to refer not to the dictionary itself, but to its key-value pairs.
For a better understanding of what is happening, I will try to give a sequence of actions and their consequences within the newly launched game.

_First the_ `cache == null`_, and once we select any thing, the cache (dictionary) is created with the first key-value pair._

![actions_table](https://user-images.githubusercontent.com/62316596/106518575-26718c80-64eb-11eb-9bcc-099cffb7d829.png)

Actions **[2-4]** - although the **tB** thing was selected after the pawn, its name displayed as expected because the pawn was deselected first.

Actions **[5-6]**, **[8-9]** and **[17-18]** - changing the target from a pawn to a thing without deselecting the former:
- **[5-6]** and **[8-9]** - we select things that have not been selected previously.
- **[17-18]** - the **tA** thing is already cached, so there is no glitch in this case.

Action **[15]** - even though there wasn't a pawn in the target first, the name is displayed glitchily because the **tC** thing was already cached.

And speaking of the game session.. If we reload this save **(but not the RW)**, the names of these things, being already in the cache, will be displayed as they were cached for the first time in this game session.
Regardless of the actions sequence, the names of things **tC** and **tE** will be displayed incorrectly.

> I'l try to look for a way to prevent this

The brusque way I see is to patch the "Truncate" method removing the first condition. ;)

**Comment by Reiquard** [comment]
Inserting a new line prior to [line 76](https://github.com/jptrrs/HumanResources/blob/1ff43c0f978f31efd12a2babd790ccc16154d29a/Source/Base/ModBaseHumanResources.cs#L76) removes the "Log" tab but also solves this issue.
```
t.inspectorTabsResolved.RemoveLast();
t.inspectorTabsResolved.Add(InspectTabManager.GetSharedInstance(typeof(ITab_PawnKnowledge)));
```
So I assume the cause of this bug is beyond your mod.
And in this case, the patch that I sent is probably not the worst solution.

**Comment by jptrrs** [comment]
> It would be more correct to refer not to the dictionary itself, but to its key-value pairs.

You are absolutely right. And knowing this, I got really curious to find out what was actually preventing the AdjustedLabelFor to get the right width for the panel. It turns out that label is determined after the ExtraOnGUI method does the tabs for the pawn, and that function actually changes the panel width to accomodate the various tabs (including my Tech tab). So, after a lot of digging in the decomplier and trying different things, I realized all it took was to intercept the GetLabel box parameter and adjust it accordingly. A one-line patch! Check out the commit.
